aws: Create IAM role for registry on S3

dcarley opened the pull request:
#### aws: Create IAM role for registry on S3

Create an IAM role and instance profile, associated with the registry
instance, which will allow it to access S3 without using a predefined
access/secret key.

The policy matches what we had previously setup by hand for the single
username/keypair. It will allow access to the buckets of all independent
environments. It's not possible to restrict this further without giving
everyone access to IAM.

NB: The `iam_instance_profile` attribute is `ForceNew` so the registry
instance will be recreated. This is OK though, because the data is in S3 and
the index will be rebuilt from S3 when the new machine comes up.
#### aws: Comment out IAM role resources

Per the comment:

> These resources are commented because not all of us have access to create
> and modify things in IAM. This has been run once by dcarley and remain
> here for reference or future modification.

Using `#` comments instead of `/* */` because hashicorp/terraform#2248
results in the following error:

```
➜  aws git:(docker_registry_s3_role) ✗ terraform apply -var env=dcarley
Error configuring: 2 error(s) occurred:

* aws_route53_record.docker-registry: missing dependency: aws_instance.docker-registry
* output.docker-registry.private_ip: missing dependency: aws_instance.docker-registry
```

dcarley committed:
aws: Create IAM role for registry on S3

Create an IAM role and instance profile, associated with the registry
instance, which will allow it to access S3 without using a predefined
access/secret key.

The policy matches what we had previously setup by hand for the single
username/keypair. It will allow access to the buckets of all independent
environments. It's not possible to restrict this further without giving
everyone access to IAM.

NB: The `iam_instance_profile` attribute is `ForceNew` so the registry
instance will be recreated. This is OK though, because the data is in S3 and
the index will be rebuilt from S3 when the new machine comes up.

dcarley committed:
aws: Comment out IAM role resources

Per the comment:

> These resources are commented because not all of us have access to create
> and modify things in IAM. This has been run once by dcarley and remain
> here for reference or future modification.

Using `#` comments instead of `/* */` because hashicorp/terraform#2248
results in the following error:

    ➜  aws git:(docker_registry_s3_role) ✗ terraform apply -var env=dcarley
    Error configuring: 2 error(s) occurred:

    * aws_route53_record.docker-registry: missing dependency: aws_instance.docker-registry
    * output.docker-registry.private_ip: missing dependency: aws_instance.docker-registry

dcarley commented:
Think I broke something.

dhilton commented:
Given I can't run the commented out sections; how would you suggest this is best tested? Run this branch in terraform against AWS, redeploying a new environment?

dcarley commented:
> Given I can't run the commented out sections; how would you suggest this is best tested? Run this branch in terraform against AWS, redeploying a new environment?

It's best tested with alphagov/tsuru-ansible#94 - deploy both branches and then add a platform or deploy an app.

jimconner commented:
I'm hitting this error when I try it.

* Error launching source instance: UnauthorizedOperation: You are not authorized to perform this operation. Encoded authorization failure message: IOHKnHw2bK44-qSojQCWHTAKUX6jlVOpYhzxceQvJDGCpmD7TwF2MGZeamLw2bPFtdppCwFF0uclO9-9MUKBzJNtK5yxjdkd9_HrqgHvBQcbO-ZZstn8zTFYrvGAYylbZAa7-Djpr61O3zVM44O-PVynKSD6jYY_aeZi-zNpc5Syf5vCEzfbymBdbzcCWE3wW4IEHKmBlQ1paLk01QvDn8o6sjS51Ck_lvgRa5eTFZH87e_pIRhgQW0rJ5flMhkJKWp2j1FkxZAXuulpBv1lJTpCiy32RNoJZcXFEevyuHpVBzJWKOMFO8RCT089QDuhO6FYucKI-54Nxd_ysnBsc6nuCK_sumKqMj4Rf1XpHjoHmCYLbTZNexvdIw_B5tkMSGBvgIhGjiT3boF1f5p5P4o
        status code: 403, request id: []

dcarley commented:
Thanks, closing while I investigate.

dcarley commented:
Fixed by adding this policy to our "contractors" group, which was based on Jim decoding the error message with `aws sts decode-authorization-message`:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "Stmt1435071087000",
            "Effect": "Allow",
            "Action": [
                "iam:PassRole"
            ],
            "Resource": [
                "arn:aws:iam::988997429095:role/tsuru-docker-registry-s3"
            ]
        }
    ]
}
```
I tested this by creating myself a new user (which I've now deleted) in that group.

jimconner commented:
It now works correctly for me...

Apply complete! Resources: 27 added, 0 changed, 0 destroyed.
